network: Fix cycle dependency causing race of netdev and address configuration

thomasfire opened the pull request:
This changes the following chain after `nixos-rebuild switch` with modified tunnel interfaces: stop network-setup -> stop TUN-netdev -> stop network-addresses-TUN -> start network-addresses-TUN (fails since it depends on TUN-netdev which is off).

Chain after this change:
stop TUN-netdev -> stop network-setup -> stop network-addresses-TUN -> start TUN-netdev -> start network-addresses-TUN -> start network-setup

Was found when we investigated https://git.m-labs.hk/M-Labs/it-infra/issues/34 .

## Reproduction steps

1. In `/etc/nixos/configuration.nix`, add any GRE tunnel and a corresponding interface addresses configuration, like here:
    ```
    greTunnels."mytunnel" = {
      dev = myEthDevice;
      remote = "X.X.X.X";
      local = "Y.Y.Y.Y";
      ttl = 255;
      type = "tun";
    };
    interfaces."mytunnel" = {
      ipv4.addresses = [
        {
          address = "Z.Z.Z.Z";
          prefixLength = N;
        }
      ];
      ipv4.routes = [
        {
          address = "0.0.0.0";
          prefixLength = 0;
          via = "R.R.R.R";
          options.table = "1";
        }
      ];
    };
   ```
2. Save and apply configuration in any way
3. Change the addresses in the `/etc/nixos/configuration.nix` again, and use `nixos-rebuild switch`
4. After patch: configuration is applied without any errors
5. Before patch: configuration fails on restarting the services with similar to the following log in journal:
    ```
    Feb 06 17:33:49 nixbld systemd[1]: network-addresses-mytunnel.service: Bound to unit sys-subsystem-net-devices-mytunnel.device, but unit isn't active.
    Feb 06 17:33:49 nixbld systemd[1]: Dependency failed for Address configuration of mytunnel.
    Feb 06 17:33:49 nixbld systemd[1]: network-addresses-alt0.service: Job network-addresses-mytunnel.service/start failed with result 'dependency'.
    Feb 06 17:35:25 nixbld systemd[1]: Dependency failed for Address configuration of mytunnel.
    Feb 06 17:35:25 nixbld systemd[1]: network-addresses-alt0.service: Job network-addresses-mytunnel.service/start failed with result 'dependency'.
    ```

## Things done


- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

@bjornfor 
@sinanmohd
@rgrunbla

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Commit:
network: Fix cycle dependency causing race of netdev and address configuration

This changes the following chain after `nixos-rebuild switch` with modified tunnel interfaces:
stop network-setup -> stop TUN-netdev -> stop network-addresses-TUN -> start network-addresses-TUN (fails since it depends on TUN-netdev which is off).

Chain after this change:
stop TUN-netdev -> stop network-setup -> stop network-addresses-TUN -> start TUN-netdev -> start network-addresses-TUN -> start network-setup

Signed-off-by: Egor Savkin <es@m-labs.hk>

rnhmjoj commented:
I think I've seen this issue in my router with this device (custom code, but copied from this module):
```nix
  systemd.services."wan4-netdev" =
    { description = "4in6 tunnel interface";
      after    = [ "network-pre.target" ];
      wantedBy = [ "network-setup.service" ];
      partOf = [ "network-setup.service" ];
      before = [ "network-setup.service" ];
      serviceConfig.Type = "oneshot";
      serviceConfig.RemainAfterExit = true;
      path = [ pkgs.iproute2 ];
      script = ''
        # set up the tunnel
        ip link show dev wan4 >/dev/null 2>&1 && ip tunnel del dev wan4
        ip tunnel add name wan4 mode ipip6 encaplimit none \
          remote ${tunnelRemote} local ${tunnelLocal}
        ip link set dev wan4 up
        ip route add default dev wan4
      '';
      postStop = "ip tunnel del wan4 || true";
    };

```
I'll try removing the PartOf and investigate a bit. Thank you!

thomasfire commented:
Hi @rnhmjoj , were you able to check it?

rnhmjoj commented:
No, sorry, but thanks for the ping: I should have time to run some tests this week.

rnhmjoj commented:
This is breaking `nixosTests.networking.scripted.virtual`:
```
Test "Test interfaces clean up" failed with error: "Some virtual interface has not been properly cleaned:
tun0: tun persist user 0
tap0: tap persist user 0
"
```
However I think the test is not correct: stopping network-addresses-* should not stop make the device disappear from `ip tuntap list`. (guess who wrote the test? https://github.com/NixOS/nixpkgs/pull/34442)

Commit:
tests: network: update `nixosTests.networking.scripted.virtual` to match correct behavior

First, `stop network-addresses-X` should clean up the addresses without deleting the interfaces. And only `stop tap0-netdev` should fully delete the interfaces.

Signed-off-by: Egor Savkin <es@m-labs.hk>

thomasfire commented:
Thank you for pointing out the tests! I updated them to match the correct behavior.

rnhmjoj commented:
Besides testing on my router, I've rebased this unto master and run more tests that use bridges and other netdevs:
- [x] nixosTests.networking.networkd
- [x] nixosTests.networking.scripted
- [x] nixosTests.containers-restart_networking -> already broken in master
- [x] nixosTests.containers-bridge
- [x] nixosTests.containers-macvlans
- [x] nixosTests.containers-extra_veth
- [x] nixosTests.containers-hosts
- [x] nixosTests.containers-physical_interfaces
- [x] nixosTests.libreswan
- [x] nixosTests.libreswan-nat
- [x] nixosTests.systemd-initrd-bridge  -> already broken in master
- [x] nixosTests.yggdrasil

So, everything seems ok.

Also, I went looking for the original commit that added these `PartOf=`s and it looks like the problem was already noticed in 2017: https://github.com/NixOS/nixpkgs/pull/18491#issuecomment-291489989. The author mentions replacing the dependecy to `network-setup.services` with `network-setup.target`, but I'm not sure what they meant because that target doesn't exist.

thomasfire commented:
Thanks for checking.

Just rebased and fixed typo `ip link` -> `ip address` in tests.

fpletz commented:
I hope we will soon ditch scripted networking and finally use networkd by default. :pray:
